reahl.component.shelltools.ExecutableNotInstalledException: Executable not found: dropdb

(reahl_env) c:\appserv\water_poll>pytest -o python_functions=demo_setup --pyargs water_poll_dev.test_water_poll
pytest -o python_functions=demo_setup --pyargs water_poll_dev.test_water_poll
============================= test session starts =============================
platform win32 -- Python 3.7.0, pytest-3.9.3, py-1.7.0, pluggy-0.8.0
rootdir: c:\appserv\water_poll, inifile:
collected 1 item

water_poll_dev\test_water_poll.py F                                    [100%]

================================== FAILURES ===================================
________ demo_setup[sql_alchemy_fixture0-water_poll_domain_fixture0] _________

self = <FunctionWrapper at 0x366b4c8 for function at 0x33e4c00>, args = ()
kwargs = {'water_poll_domain_fixture': WaterPollDomainFixture[default_scenario], 'sql_alchemy_fixture': SqlAlchemyFixture[default_scenario]}
instance = None

    def __call__(self, *args, **kwargs):
        # If enabled has been specified, then evaluate it at this point
        # and if the wrapper is not to be executed, then simply return
        # the bound function rather than a bound wrapper for the bound
        # function. When evaluating enabled, if it is callable we call
        # it, otherwise we evaluate it as a boolean.
    
        if self._self_enabled is not None:
            if callable(self._self_enabled):
                if not self._self_enabled():
                    return self.__wrapped__(*args, **kwargs)
            elif not self._self_enabled:
                return self.__wrapped__(*args, **kwargs)
    
        # This can occur where initial function wrapper was applied to
        # a function that was already bound to an instance. In that case
        # we want to extract the instance from the function and use it.
    
        if self._self_binding == 'function':
            if self._self_instance is None:
                instance = getattr(self.__wrapped__, '__self__', None)
                if instance is not None:
                    return self._self_wrapper(self.__wrapped__, instance,
                            args, kwargs)
    
        # This is generally invoked when the wrapped function is being
        # called as a normal function and is not bound to a class as an
        # instance method. This is also invoked in the case where the
        # wrapped function was a method, but this wrapper was in turn
        # wrapped using the staticmethod decorator.
    
        return self._self_wrapper(self.__wrapped__, self._self_instance,
>               args, kwargs)

c:\prg\reahl_env\lib\site-packages\wrapt\wrappers.py:523: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
c:\prg\reahl_env\lib\site-packages\reahl\tofu\pytestsupport.py:152: in test_with_fixtures
    stack.enter_context(fixture)
c:\development\Lib\contextlib.py:426: in enter_context
    result = _cm_type.__enter__(cm)
c:\prg\reahl_env\lib\site-packages\reahl\dev\fixtures.py:55: in __enter__
    return super(ContextAwareFixture, self).__enter__()
c:\prg\reahl_env\lib\site-packages\reahl\tofu\fixture.py:320: in __enter__
    self.run_marked_methods(SetUp, order=reversed)
c:\prg\reahl_env\lib\site-packages\reahl\tofu\fixture.py:291: in run_marked_methods
    marked.method_for(self)()
c:\prg\reahl_env\lib\site-packages\reahl\dev\fixtures.py:127: in init_database
    self.system_control.initialise_database(yes=True)
c:\prg\reahl_env\lib\site-packages\reahl\component\dbutils.py:152: in initialise_database
    self.drop_database(yes=yes)
c:\prg\reahl_env\lib\site-packages\reahl\component\dbutils.py:185: in drop_database
    return self.db_control.drop_database(super_user_name=super_user_name, yes=yes)
c:\prg\reahl_env\lib\site-packages\reahl\postgresqlsupport.py:75: in drop_database
    Executable('dropdb').check_call(cmd_args + last_part)
c:\prg\reahl_env\lib\site-packages\reahl\component\shelltools.py:71: in check_call
    return self.execute(subprocess.check_call, commandline_arguments, *args, **kwargs)
c:\prg\reahl_env\lib\site-packages\reahl\component\shelltools.py:66: in execute
    return method([self.executable_file]+commandline_arguments, *args, **kwargs)
c:\prg\reahl_env\lib\site-packages\reahl\component\shelltools.py:52: in executable_file
    return self.which(self.name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <reahl.component.shelltools.Executable object at 0x05568AD0>
program = 'dropdb'

    def which(self, program):
        #on windows os, some entrypoints installed in the virtualenv
        #need their full path(with extension) to be able to be used as a spawn command.
        #Python 3 now offers shutil.which() - see also http://stackoverflow.com/questions/377017/test-if-executable-exists-in-python
        executable = distutils.spawn.find_executable(program)
        if executable:
            return executable
>       raise ExecutableNotInstalledException(program)
E       reahl.component.shelltools.ExecutableNotInstalledException: Executable not found: dropdb

c:\prg\reahl_env\lib\site-packages\reahl\component\shelltools.py:61: ExecutableNotInstalledException
------------------------------ Captured log call ------------------------------
config.py                  198 WARNING  web.static_root in stage/2-sandbox\web.config.py is using a dangerous default setting: c:\appserv\water_poll
============================== warnings summary

--- a/reahl-component/reahl/component/exceptions.py
+++ b/reahl-component/reahl/component/exceptions.py
@@ -34,6 +34,7 @@ else:
 
 _ = Catalogue('reahl-component')
 
+
 class DomainException(Exception):
     """Any exception indicating an application-specific error condition that 
        should be communicated to a user.

--- a/reahl-postgresqlsupport/reahl/postgresqlsupport.py
+++ b/reahl-postgresqlsupport/reahl/postgresqlsupport.py
@@ -33,12 +33,21 @@ import io
 import subprocess
 import gzip
 from datetime import date
-from contextlib import closing
+from contextlib import closing, contextmanager
 import os.path
 import socket
 
 from reahl.component.dbutils import DatabaseControl
-from reahl.component.shelltools import Executable
+from reahl.component.exceptions import DomainException
+from reahl.component.shelltools import Executable, ExecutableNotInstalledException
+
+
+@contextmanager
+def reraise_ExecutableNotInstalled_as_DomainException():
+    try:
+        yield
+    except ExecutableNotInstalledException as e:
+        raise DomainException(message=six.text_type(e))
 
 
 class PostgresqlControl(DatabaseControl):
@@ -57,13 +66,15 @@ class PostgresqlControl(DatabaseControl):
 
     def create_db_user(self, super_user_name=None, create_with_password=True):
         create_password_option = 'P' if create_with_password else ''
-        Executable('createuser').check_call(['-DSRl%s' % create_password_option]
+        with reraise_ExecutableNotInstalled_as_DomainException():
+            Executable('createuser').check_call(['-DSRl%s' % create_password_option]
                                             + self.login_args(login_username=super_user_name)
                                             + [self.user_name])
         return 0
 
     def drop_db_user(self, super_user_name=None):
-        Executable('dropuser').check_call(self.login_args(login_username=super_user_name) + [self.user_name])
+        with reraise_ExecutableNotInstalled_as_DomainException():
+            Executable('dropuser').check_call(self.login_args(login_username=super_user_name) + [self.user_name])
         return 0
 
     def drop_database(self, super_user_name=None, yes=False):
@@ -72,12 +83,14 @@ class PostgresqlControl(DatabaseControl):
         if yes:
             last_part = [self.database_name]
 
-        Executable('dropdb').check_call(cmd_args + last_part)
+        with reraise_ExecutableNotInstalled_as_DomainException():
+            Executable('dropdb').check_call(cmd_args + last_part)
         return 0
 
     def create_database(self, super_user_name=None):
         owner_option = ['-O', self.user_name] if self.user_name else []
-        Executable('createdb').check_call(['-Eunicode']
+        with reraise_ExecutableNotInstalled_as_DomainException():
+            Executable('createdb').check_call(['-Eunicode']
                                           + self.login_args(login_username=super_user_name)
                                           + ['-T', 'template0']
                                           + owner_option + [self.database_name])
@@ -87,7 +100,8 @@ class PostgresqlControl(DatabaseControl):
         today = date.today()
         filename = '%s.psql.%s' % (self.database_name, today.strftime('%A'))
         full_path = os.path.join(directory, filename)
-        with io.open(full_path, 'w') as destination_file:
+        with io.open(full_path, 'w') as destination_file,\
+                reraise_ExecutableNotInstalled_as_DomainException():
             cmd_args = ['-Fc', '-o'] + self.login_args(login_username=super_user_name) + [self.database_name]
             Executable('pg_dump').check_call(cmd_args, stdout=destination_file)
         return 0
@@ -100,21 +114,23 @@ class PostgresqlControl(DatabaseControl):
         filename = '%s-all.%s.sql.gz' % (hostname, today.strftime('%A'))
         full_path = os.path.join(directory, filename)
 
-        with closing(gzip.open(full_path, 'wb')) as zipped_file:
+        with closing(gzip.open(full_path, 'wb')) as zipped_file, \
+                reraise_ExecutableNotInstalled_as_DomainException():
             proc = Executable('pg_dumpall').Popen(['-o'] + self.login_args(login_username=super_user_name),
                                                   stdout=subprocess.PIPE)
             for line in proc.stdout:
                 zipped_file.write(line)
         return 0
 
-
     def restore_database(self, filename, super_user_name=None):
-        Executable('pg_restore').check_call(self.login_args(login_username=super_user_name)
+        with reraise_ExecutableNotInstalled_as_DomainException():
+            Executable('pg_restore').check_call(self.login_args(login_username=super_user_name)
                                             + ['-C', '-Fc', '-d', 'postgres', filename])
         return 0
 
     def restore_all_databases(self, filename, super_user_name=None):
-        with closing(gzip.open(filename, 'rb')) as zipped_file:
+        with closing(gzip.open(filename, 'rb')) as zipped_file, \
+                reraise_ExecutableNotInstalled_as_DomainException():
             proc = Executable('psql').Popen(self.login_args(login_username=super_user_name)
                                             + ['-d', 'postgres'], stdin=subprocess.PIPE)
             for line in zipped_file:
@@ -125,3 +141,4 @@ class PostgresqlControl(DatabaseControl):
         sql = 'select pg_size_pretty(pg_database_size(\'%s\'));' % self.database_name
         result = orm_control.execute_one(sql)
         return result[0]
+

--- a/reahl-postgresqlsupport/reahl/postgresqlsupport.py
+++ b/reahl-postgresqlsupport/reahl/postgresqlsupport.py
@@ -43,10 +43,10 @@ from reahl.component.shelltools import Executable, ExecutableNotInstalledExcepti
 
 
 @contextmanager
-def reraise_ExecutableNotInstalled_as_DomainException():
+def as_domain_exception(exception_to_translate):
     try:
         yield
-    except ExecutableNotInstalledException as e:
+    except exception_to_translate as e:
         raise DomainException(message=six.text_type(e))
 
 
@@ -66,14 +66,14 @@ class PostgresqlControl(DatabaseControl):
 
     def create_db_user(self, super_user_name=None, create_with_password=True):
         create_password_option = 'P' if create_with_password else ''
-        with reraise_ExecutableNotInstalled_as_DomainException():
+        with as_domain_exception(ExecutableNotInstalledException):
             Executable('createuser').check_call(['-DSRl%s' % create_password_option]
-                                            + self.login_args(login_username=super_user_name)
-                                            + [self.user_name])
+                                                + self.login_args(login_username=super_user_name)
+                                                + [self.user_name])
         return 0
 
     def drop_db_user(self, super_user_name=None):
-        with reraise_ExecutableNotInstalled_as_DomainException():
+        with as_domain_exception(ExecutableNotInstalledException):
             Executable('dropuser').check_call(self.login_args(login_username=super_user_name) + [self.user_name])
         return 0
 
@@ -83,27 +83,27 @@ class PostgresqlControl(DatabaseControl):
         if yes:
             last_part = [self.database_name]
 
-        with reraise_ExecutableNotInstalled_as_DomainException():
+        with as_domain_exception(ExecutableNotInstalledException):
             Executable('dropdb').check_call(cmd_args + last_part)
         return 0
 
     def create_database(self, super_user_name=None):
         owner_option = ['-O', self.user_name] if self.user_name else []
-        with reraise_ExecutableNotInstalled_as_DomainException():
+        with as_domain_exception(ExecutableNotInstalledException):
             Executable('createdb').check_call(['-Eunicode']
-                                          + self.login_args(login_username=super_user_name)
-                                          + ['-T', 'template0']
-                                          + owner_option + [self.database_name])
+                                              + self.login_args(login_username=super_user_name)
+                                              + ['-T', 'template0']
+                                              + owner_option + [self.database_name])
         return 0
 
     def backup_database(self, directory, super_user_name=None):
         today = date.today()
         filename = '%s.psql.%s' % (self.database_name, today.strftime('%A'))
         full_path = os.path.join(directory, filename)
-        with io.open(full_path, 'w') as destination_file,\
-                reraise_ExecutableNotInstalled_as_DomainException():
+        with io.open(full_path, 'w') as destination_file:
             cmd_args = ['-Fc', '-o'] + self.login_args(login_username=super_user_name) + [self.database_name]
-            Executable('pg_dump').check_call(cmd_args, stdout=destination_file)
+            with as_domain_exception(ExecutableNotInstalledException):
+                Executable('pg_dump').check_call(cmd_args, stdout=destination_file)
         return 0
 
     def backup_all_databases(self, directory, super_user_name=None):
@@ -114,25 +114,25 @@ class PostgresqlControl(DatabaseControl):
         filename = '%s-all.%s.sql.gz' % (hostname, today.strftime('%A'))
         full_path = os.path.join(directory, filename)
 
-        with closing(gzip.open(full_path, 'wb')) as zipped_file, \
-                reraise_ExecutableNotInstalled_as_DomainException():
-            proc = Executable('pg_dumpall').Popen(['-o'] + self.login_args(login_username=super_user_name),
-                                                  stdout=subprocess.PIPE)
+        with closing(gzip.open(full_path, 'wb')) as zipped_file:
+            with as_domain_exception(ExecutableNotInstalledException):
+                proc = Executable('pg_dumpall').Popen(['-o'] + self.login_args(login_username=super_user_name),
+                                                      stdout=subprocess.PIPE)
             for line in proc.stdout:
                 zipped_file.write(line)
         return 0
 
     def restore_database(self, filename, super_user_name=None):
-        with reraise_ExecutableNotInstalled_as_DomainException():
+        with as_domain_exception(ExecutableNotInstalledException):
             Executable('pg_restore').check_call(self.login_args(login_username=super_user_name)
                                             + ['-C', '-Fc', '-d', 'postgres', filename])
         return 0
 
     def restore_all_databases(self, filename, super_user_name=None):
-        with closing(gzip.open(filename, 'rb')) as zipped_file, \
-                reraise_ExecutableNotInstalled_as_DomainException():
-            proc = Executable('psql').Popen(self.login_args(login_username=super_user_name)
-                                            + ['-d', 'postgres'], stdin=subprocess.PIPE)
+        with closing(gzip.open(filename, 'rb')) as zipped_file:
+            with as_domain_exception(ExecutableNotInstalledException):
+                proc = Executable('psql').Popen(self.login_args(login_username=super_user_name)
+                                                + ['-d', 'postgres'], stdin=subprocess.PIPE)
             for line in zipped_file:
                 proc.stdin.write(line)
         return 0

